crates/polars-utils/src/mmap.rs

```diff
@@ -1,90 +1,145 @@
-    position: usize,
-}
+    impl Deref for MemSlice {
+        type Target = [u8];
 
-/// A read-only reference to a slice of memory.
-///
-/// This memory can either be heap-allocated or be mmap-ed into memory.
-///
-/// This still owns the all the original memory and therefore should probably not be a long-lasting
-/// structure.
-#[derive(Clone)]
-pub struct MemSlice(MemSliceInner);
-
-#[derive(Clone)]
-enum MemSliceInner {
-    Mmap(MmapSlice),
-    Allocated(AllocatedSlice),
-}
+        #[inline(always)]
+        fn deref(&self) -> &Self::Target {
+            self.slice
+        }
+    }
 
-#[derive(Clone)]
-struct AllocatedSlice {
-    data: Arc<[u8]>,
-    start: usize,
-    end: usize,
-}
+    impl Default for MemSlice {
+        fn default() -> Self {
+            Self {
+                slice: &[],
+                inner: MemSliceInner::Unbacked,
+            }
+        }
+    }
 
-impl Deref for MmapSlice {
-    type Target = [u8];
+    impl MemSlice {
+        /// Copy the contents into a new owned `Vec`
+        #[inline(always)]
+        pub fn to_vec(self) -> Vec<u8> {
+            <[u8]>::to_vec(self.deref())
+        }
 
-    fn deref(&self) -> &Self::Target {
-        // SAFETY: Invariant of MmapSlice
-        unsafe { std::slice::from_raw_parts(self.ptr, self.len) }
-    }
-}
+        /// Construct a `MemSlice` from an existing `Vec<u8>`. This is zero-copy.
+        #[inline]
+        pub fn from_vec(v: Vec<u8>) -> Self {
+            let arc_vec = Arc::new(v);
 
-impl Deref for MemSlice {
-    type Target = [u8];
+            Self {
+                slice: unsafe { std::mem::transmute::<&[u8], &'static [u8]>(arc_vec.as_slice()) },
+                inner: MemSliceInner::Vec(arc_vec),
+            }
+        }
 
-    #[inline(always)]
-    fn deref(&self) -> &Self::Target {
-        match &self.0 {
-            MemSliceInner::Mmap(v) => v.deref(),
-            MemSliceInner::Allocated(v) => &v.data[v.start..v.end],
+        /// Construct a `MemSlice` from [`bytes::Bytes`]. This is zero-copy.
+        #[inline]
+        pub fn from_bytes(bytes: bytes::Bytes) -> Self {
+            Self {
+                slice: unsafe { std::mem::transmute::<&[u8], &'static [u8]>(bytes.as_ref()) },
+                inner: MemSliceInner::Bytes(bytes),
+            }
         }
-    }
-}
 
-impl Default for AllocatedSlice {
-    fn default() -> Self {
-        let slice: &[u8] = &[];
-        Self {
-            data: Arc::from(slice),
-            start: 0,
-            end: 0,
+        #[inline]
+        pub fn from_mmap(mmap: Arc<Mmap>) -> Self {
+            Self {
+                slice: unsafe {
+                    std::mem::transmute::<&[u8], &'static [u8]>(mmap.as_ref().as_ref())
+                },
+                inner: MemSliceInner::Mmap(mmap),
+            }
+        }
+
+        /// Construct a `MemSlice` that simply wraps around a `&[u8]`.
+        #[inline]
+        pub fn from_slice(slice: &'static [u8]) -> Self {
```

Comment:
Changes `from_slice` - previously it copied into a new `Arc<[u8]>`, but now it simply wraps the slice.

For context, this is only used by `ReaderBytes<'a>`. I think one day we can refactor/relace `ReaderBytes` - the way we currently use that `'a` lifetime is not ideal.